File: practicas/practica4_rb/lib/SpecificDamage.rb
#encoding: utf-8

require_relative 'Damage'
require_relative 'SpecificDamageToUI'

module Deepspace
  class SpecificDamage < Damage
    public_class_method :new

    attr_reader :weapons, :nShields

    def initialize(w, s)
      @weapons = w.clone
      super s
    end

    def self.newCopy(d)
      new(d.weapons, d.nShields)
    end

    def discardWeapon(w)
      indice = weapons.index(w.type)
      if indice != nil
        @weapons.delete_at(indice)
      end
    end

    def hasNoEffect
      @weapons.empty? and super
    end

    private def arrayContainsType(w, t)
      indice = w.index { |x| x.type == t }
      if indice != nil
        indice
      else
        -1
      end
    end

    def getUIversion
      SpecificDamageToUI.new(self)
    end


    def copy
      SpecificDamage.newCopy self
    end

    def to_s
      "weapons: " + @weapons.to_s + "\n" + super
    end

    def adjust(w, s)
      result = []
      w_aux = w.clone
      weapons.each do |element|
        indice = arrayContainsType(w_aux, element)

        if indice != -1
          result.push(element)
          w_aux.delete_at(indice)
        end
      end

      SpecificDamage.new(result, adjustShields(s))
    end
  end
end

File: practicas/practica4/src/main/java/deepspace/SpecificDamage.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author inmagalvez
 */
public class SpecificDamage extends Damage {
    
   private ArrayList<WeaponType> weapons;
   
   SpecificDamage(ArrayList<WeaponType> wl, int s) {
        super(s);
        weapons = new ArrayList<>(wl);
    }
   
   SpecificDamage(SpecificDamage other){
        this(other.weapons, other.getNShields());
    }
    
    @Override
    Damage copy(){
        return new SpecificDamage(this);
    }
    
    private int arrayContainsType(ArrayList<Weapon> w, WeaponType t) {
        int index = -1;
        for (int i = 0; i < w.size(); i++) {
            if(w.get(i).getType() == t) {
                index = i;
                break;
            }
        }
        
        return index;
    }
    
    @Override
    public Damage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s) {
        int shields = adjustShields(s);

        ArrayList<Weapon> wAux = new ArrayList<>(w);
        ArrayList<WeaponType> toSet = new ArrayList<>();

        for(WeaponType element: weapons) {
           int index = arrayContainsType(wAux, element);

            if(index != -1){  //If the element is found
                toSet.add(element); //It can be removed
                wAux.remove(index);
            }
        }

        return new SpecificDamage(toSet, shields);
    }
    
    @Override
    public void discardWeapon(Weapon w) {
        
        int index = weapons.indexOf(w.getType());
        if(index != -1){
            weapons.remove(index);
        }
        
    }
   
    
    @Override
    public boolean hasNoEffect() {
        return (weapons.isEmpty() && super.hasNoEffect());
    }
    
    public ArrayList<WeaponType> getWeapons() {
        if(weapons==null){
            return null;
        }
        return new ArrayList<>(weapons);
    }
    
    
    @Override
    public String toString() {
        return super.toString() + "\nweapons: " + getWeapons();
    }
    
    @Override
    DamageToUI getUIversion() {
        return new SpecificDamageToUI(this);
    }
    
    
}
